### DISCUSSION TITLE
Are there any example repos that show how to build on top of this?

### DISCUSSION THREAD

**Discussion by lexi-brt** [ros2-rust/ros2_rust#271] [Q&A]
I want to start building some ros2 nodes in rust, but it would be really helpful to see another project that has examples of how to set it all up. I know there are examples in this repo, but it would be more helpful  to see an entirely separate project so that it would be more clear how dependencies on this are linked in.

**Comment by nnmm** [comment]
Hi lexi-brt, we'll move the examples over to a new example repo soon, but basically the setup is similar to the one described here: https://github.com/ros2-rust/ros2_rust#sounds-great-how-can-i-try-this-out

That is, you need to:
- have dependencies installed (ROS 2, clang, cargo-ament-build, colcon-cargo, colcon-ros-cargo)
- add the https://github.com/ros2/rosidl_defaults.git repo, the `rosidl_generator_rs` package from ros2-rust and all message packages you want to build to your workspace (this will hopefully not be necessary anymore in the future)
- source the ROS 2 installation
- run `colcon build`

This will build using the `rclrs` version published on crates.io. If you clone `rclrs` into your workspace, it will use that one instead – useful if you want to use features that have not yet landed in the crates.io releases.

If you try it out and you encounter any obstacles, let me know!

**Comment by keenanjohnson** [comment]
Hey @lexi-brt ! I had the same question / desire as you when I stumbled accross this project, so I created this: https://github.com/keenanjohnson/ros2_rust_workspace

It's not perfect and I'm sure there are lots of improvements you can see, but hopefully it's helpful to you as something of a starting place that shows how you would pull in the ros2_rust repo as a submodule, and build your own workspace, etc.

**Comment by lexi-brt** [comment]
@keenanjohnson This is awesome! Thank you!

Does ros2_rust have to live in the same workspace as my package? Is there a way to keep it in it's own space and source it in the same way that you source foxy before building your own workspace?

**Reply by nnmm** [reply]
Yes, you can check out ros2_rust somewhere else, build it, and `source install/setup.bash`, then continue working in your workspace.

**Reply by keenanjohnson** [reply]
@lexi-brt if keeping ros2_rust in a seperate workspace as part of the docker file in the project I linked above makes more sense from your perspective, I think that could be a great contribution you could make if interested to https://github.com/keenanjohnson/ros2_rust_workspace
